app/src/main/java/de/westnordost/streetcomplete/data/osm/upload/SingleOsmElementTagChangesUpload.kt

```diff
@@ -0,0 +1,129 @@
+                }
+                handlingConflict = true
+                element = handleConflict(quest, element, e)
+                // try again (go back to beginning of loop)
+                continue
+            }
+            return handler.updatedElements.single()
+        }
+    }
+
+    private fun handleConflict(quest: HasElementTagChanges, element: Element, e: OsmConflictException): Element {
+        /* Conflict can either happen because of the changeset or because of the element(s)
+           uploaded. A changeset conflict cannot be handled here */
+        val newElement = element.fetchUpdated()
+        if (newElement?.version == element.version) throw ChangesetConflictException(e.message, e)
+
+        if (newElement == null) {
+            throw ElementDeletedException("Element has already been deleted")
+        }
+
+        if (isGeometrySubstantiallyDifferent(element, newElement)) {
+            throw ElementIncompatibleException("Element geometry changed substantially")
+        }
+
+        /* if after updating to the new version of the element, the quest is not applicable to
+           the element anymore, drop it (#720) */
+        if (quest.isApplicableTo(newElement) == false) {
+            throw ElementConflictException("Quest no longer applicable to the element")
+        }
+        return newElement
+    }
+
+    private fun Element.fetchUpdated() =
+        when (this) {
+            is Node -> osmDao.getNode(id)
+            is Way -> osmDao.getWay(id)
+            is Relation -> osmDao.getRelation(id)
+            else -> null
+        }
+}
+
+private fun isGeometrySubstantiallyDifferent(element: Element, newElement: Element) =
+    when (element) {
+        is Node -> isNodeGeometrySubstantiallyDifferent(element, newElement as Node)
+        is Way -> isWayGeometrySubstantiallyDifferent(element, newElement as Way)
+        is Relation -> isRelationGeometrySubstantiallyDifferent(element, newElement as Relation)
+        else -> false
+    }
+
+private fun isNodeGeometrySubstantiallyDifferent(node: Node, newNode: Node) =
+    /* Moving the node a distance beyond what would pass as adjusting the position within a
+       building counts as substantial change. Also, the maximum distance should be not (much)
+       bigger than the usual GPS inaccuracy in the city. */
+    SphericalEarthMath.distance(node.position, newNode.position) > 20
+
+private fun isWayGeometrySubstantiallyDifferent(way: Way, newWay: Way) =
+    /* if the first or last node is different, it means that the way has either been extended or
+       shortened at one end, which is counted as being substantial:
+       If for example the surveyor has been asked to determine something for a certain way
+       and this way is now longer, his answer does not apply to the whole way anymore, so that
+       is an unsolvable conflict. */
+    way.nodeIds.firstOrNull() != newWay.nodeIds.firstOrNull() ||
```

Comment:
What about entire way being moved? I made/notices such edits where objects were dragged for a long distance rather than delete/recreate it.

Reply:
This kind of reusing element ids for entirely different things is frowned upon in the community, because it carries over the history of a node that is entirely unrelated to its current usage. Additionally, the moving of its nodes does not appear in the history of the way or relation, making it really hard to retrace the "real" history in a case like this. In short, just don't do it.
So, in the unlikely event that a mistake is introduced because (most likely) a node got this treatment, I don't think anyone would blame StreetComplete for it. 

If StreetComplete were to safeguard against this, it would mean that any even so slight adjustment of the position of the node and worse, any node within a way, would need to be treated as an unsolvable conflict. (Fuzzy "if X% of nodes moved more than Ym"-checks are out of the question) Slight adjustments of nodes are very common.

In short, I won't make this check more strict.